### PR TITLE
feat: channel upgradability

### PR DESCRIPTION
**Pull request by damiannolan**
## Description

Placeholder PR and feature branch for `04-channel` upgrades

closes: #1599 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: .github/compatibility-test-matrices/unreleased/transfer.json
Comment (damiannolan) [inline .github/compatibility-test-matrices/unreleased/transfer.json:1]: this file was kept around from merge. I'll remove it. This file is now split up on main

**Comment by damiannolan** [comment]
will fix linter shortly

**Comment by damiannolan** [comment]
It looks like https://github.com/cosmos/ibc-go/pull/1613/commits/fa3f92fbb546cc7f450f796eaac14bfc21d60022 did not merge correctly and the github UI was still rendering 200+ files changed after merge - I think this was due to using a squash merge commit. 

@chatton and I merged main back in here in https://github.com/cosmos/ibc-go/pull/1613/commits/b41b510e129d15ca07acc30a9760c6bafa5738ea and looks to be fine now.

File: e2e/sample.config.yaml [hunk 1]
```diff
@@ -30,7 +30,7 @@ relayers:
     tag: "latest"
   - id: hyperspace
     image: ghcr.io/misko9/hyperspace
-    tag: "timeout"
+    tag: "20231122v39"
```
Comment (crodriguezvega) [inline e2e/sample.config.yaml:33]: This was updated to have the happy path e2e working.

File: e2e/tests/wasm/grandpa_test.go [hunk 2]
```diff
@@ -754,12 +754,12 @@ func (s *GrandpaTestSuite) GetGrandpaTestChains() (ibc.Chain, ibc.Chain) {
...
 			{
 				Repository: "ghcr.io/misko9/polkadot-node",
-				Version:    "local",
+				Version:    "v39",
 				UidGid:     "1000:1000",
 			},
 			{
 				Repository: "ghcr.io/misko9/parachain-node",
-				Version:    "latest",
+				Version:    "20231122v39",
```
Comment (crodriguezvega) [inline e2e/tests/wasm/grandpa_test.go:762]: Also needed to be updated for the e2e to pass.

File: modules/core/03-connection/keeper/verify.go [hunk 3]
```diff
@@ -366,6 +366,127 @@ func (k Keeper) VerifyNextSequenceRecv(
...
+	); err != nil {
+		return errorsmod.Wrapf(err, "failed upgrade error receipt verification for client (%s)", clientID)
+	}
+
+	return nil
+}
+
+// VerifyChannelUpgradeErrorAbsence verifies a proof of the absence of a
+// channel upgrade error.
+func (k Keeper) VerifyChannelUpgradeErrorAbsence(
```
Comment (damiannolan) [inline modules/core/03-connection/keeper/verify.go:413]: I don't think this is used anymore?

**Review by crodriguezvega — APPROVED** [review]
CI is passing and overall it looks good, so throwing my approval. :) Phenomenal work everyone!

File: modules/core/03-connection/keeper/verify.go (hunk 3 above)
Comment (crodriguezvega) [reply modules/core/03-connection/keeper/verify.go:413]: Just checked and you're right!

File: modules/apps/27-interchain-accounts/host/ibc_module.go [hunk 4]
```diff
@@ -156,6 +157,32 @@ func (IBCModule) OnTimeoutPacket(
...
+
+// OnChanUpgradeAck implements the IBCModule interface
+func (IBCModule) OnChanUpgradeAck(ctx sdk.Context, portID, channelID, counterpartyVersion string) error {
+	return errorsmod.Wrap(icatypes.ErrInvalidChannelFlow, "channel upgrade handshake must be initiated by controller chain")
+}
+
+// OnChanUpgradeOpen implements the IBCModule interface
+func (IBCModule) OnChanUpgradeOpen(ctx sdk.Context, portID, channelID string, order channeltypes.Order, connectionHops []string, version string) {
+}
+
```
Comment (charleenfei) [inline modules/apps/27-interchain-accounts/host/ibc_module.go:182]: pass thru here

**Review by chatton — APPROVED** [review]
I did not look at test code, or dig into implementation specifics (we did all that in the audits as well as PR reviews! )

I mainly checked to make sure everything was in the right place. 

CI is passing so LGTM

File: modules/apps/29-fee/ibc_middleware.go [hunk 5]
```diff
@@ -326,6 +327,129 @@ func (im IBCMiddleware) OnTimeoutPacket(
...
+		return
+	}
+
+	// set fee enabled and passthrough to the next middleware of application in callstack.
+	im.keeper.SetFeeEnabled(ctx, portID, channelID)
+	cbs.OnChanUpgradeOpen(ctx, portID, channelID, order, connectionHops, versionMetadata.AppVersion)
+}
+
+// OnChanUpgradeRestore implements the IBCModule interface
+func (im IBCMiddleware) OnChanUpgradeRestore(ctx sdk.Context, portID, channelID string) {
```
Comment (colin-axner) [inline modules/apps/29-fee/ibc_middleware.go:444]: no test

File: modules/apps/27-interchain-accounts/host/ibc_module.go [hunk 6]
```diff
@@ -156,6 +157,32 @@ func (IBCModule) OnTimeoutPacket(
 	return errorsmod.Wrap(icatypes.ErrInvalidChannelFlow, "cannot cause a packet timeout on a host channel end, a host chain does not send a packet over the channel")
 }
 
+// OnChanUpgradeInit implements the IBCModule interface
+func (IBCModule) OnChanUpgradeInit(ctx sdk.Context, portID, channelID string, order channeltypes.Order, connectionHops []string, version string) (string, error) {
+	return "", errorsmod.Wrap(icatypes.ErrInvalidChannelFlow, "channel upgrade handshake must be initiated by controller chain")
```
Comment (colin-axner) [inline modules/apps/27-interchain-accounts/host/ibc_module.go:162]: no test

File: modules/apps/27-interchain-accounts/host/ibc_module.go [hunk 7]
```diff
@@ -156,6 +157,32 @@ func (IBCModule) OnTimeoutPacket(
...
+	if !im.keeper.GetParams(ctx).HostEnabled {
+		return "", types.ErrHostSubModuleDisabled
+	}
+
+	return im.keeper.OnChanUpgradeTry(ctx, portID, channelID, order, connectionHops, counterpartyVersion)
+}
+
+// OnChanUpgradeAck implements the IBCModule interface
+func (IBCModule) OnChanUpgradeAck(ctx sdk.Context, portID, channelID, counterpartyVersion string) error {
+	return errorsmod.Wrap(icatypes.ErrInvalidChannelFlow, "channel upgrade handshake must be initiated by controller chain")
```
Comment (colin-axner) [inline modules/apps/27-interchain-accounts/host/ibc_module.go:176]: no test

File: modules/apps/27-interchain-accounts/controller/ibc_middleware.go [hunk 8]
```diff
@@ -231,6 +232,100 @@ func (im IBCMiddleware) OnTimeoutPacket(
...
+			return "", errorsmod.Wrap(porttypes.ErrInvalidRoute, "upgrade route not found to module in application callstack")
+		}
+		return cbs.OnChanUpgradeInit(ctx, portID, channelID, order, connectionHops, version)
+	}
+
+	return version, nil
+}
+
+// OnChanUpgradeTry implements the IBCModule interface
+func (IBCMiddleware) OnChanUpgradeTry(ctx sdk.Context, portID, channelID string, order channeltypes.Order, connectionHops []string, counterpartyVersion string) (string, error) {
```
Comment (colin-axner) [inline modules/apps/27-interchain-accounts/controller/ibc_middleware.go:264]: no test

**Review by charleenfei — APPROVED** [review]
trusting the ci gods

File: modules/core/04-channel/keeper/upgrade.go [hunk 9]
```diff
@@ -0,0 +1,1015 @@
...
+		return err
+	}
+
+	// proof must be from a height after timeout has elapsed. Either timeoutHeight or timeoutTimestamp must be defined.
+	// if timeoutHeight is defined and proof is from before timeout height, abort transaction
+	timeoutHeight := upgrade.Timeout.Height
+	timeoutTimeStamp := upgrade.Timeout.Timestamp
+	if (timeoutHeight.IsZero() || proofHeight.LT(timeoutHeight)) &&
+		(timeoutTimeStamp == 0 || proofTimestamp < timeoutTimeStamp) {
+		return errorsmod.Wrap(types.ErrInvalidUpgradeTimeout, "upgrade timeout has not been reached for height or timestamp")
```
Comment (colin-axner) [inline modules/core/04-channel/keeper/upgrade.go:735]: can use timeout type here

File: modules/core/03-connection/keeper/verify.go (hunk 3 above)
Comment (damiannolan) [reply modules/core/03-connection/keeper/verify.go:413]: https://github.com/cosmos/ibc-go/issues/5477

File: modules/core/04-channel/keeper/upgrade.go [hunk 10]
```diff
@@ -0,0 +1,1015 @@
...
+	// delete state associated with upgrade which is no longer required.
+	k.deleteUpgradeInfo(ctx, portID, channelID)
+
+	k.SetUpgradeErrorReceipt(ctx, portID, channelID, err.GetErrorReceipt())
+
+	return channel
+}
+
+// writeErrorReceipt will write an error receipt from the provided UpgradeError.
+func (k Keeper) writeErrorReceipt(ctx sdk.Context, portID, channelID string, upgradeError *types.UpgradeError) {
```
Comment (colin-axner) [inline modules/core/04-channel/keeper/upgrade.go:1000]: looks like this function is unused? :eyes:

File: modules/core/04-channel/keeper/upgrade.go [hunk 11]
```diff
@@ -0,0 +1,1015 @@
...
+	}
+
+	// the channel upgrade sequence has already been updated in ChannelUpgradeTry, so we can pass
+	// its updated value.
+	k.restoreChannel(ctx, portID, channelID, channel.UpgradeSequence, channel, upgradeError)
+	return nil
+}
+
+// restoreChannel will restore the channel state to its pre-upgrade state so that upgrade is aborted.
+func (k Keeper) restoreChannel(ctx sdk.Context, portID, channelID string, upgradeSequence uint64, channel types.Channel, err *types.UpgradeError) types.Channel {
```
Comment (AdityaSripal) [inline modules/core/04-channel/keeper/upgrade.go:985]: Do we ever use a sequence other than channel sequence here?

**Review by AdityaSripal — APPROVED** [review]
Excellent work everyone!!! 🎉 🎉 🎉 

All the logic looks good to me. I saw some functions/args that seem unused.

There were some tests that could have additional checks, but I saw they were tested in later functions so I'm ok leaving as-is

File: modules/core/04-channel/keeper/upgrade_test.go [hunk 12]
```diff
@@ -0,0 +1,2892 @@
...
+
+			tc.malleate()
+
+			upgrade, err := suite.chainA.GetSimApp().IBCKeeper.ChannelKeeper.ChanUpgradeInit(
+				suite.chainA.GetContext(), path.EndpointA.ChannelConfig.PortID, path.EndpointA.ChannelID, upgradeFields,
+			)
+
+			if tc.expPass {
+				ctx := suite.chainA.GetContext()
+				suite.chainA.GetSimApp().IBCKeeper.ChannelKeeper.WriteUpgradeInitChannel(ctx, path.EndpointA.ChannelConfig.PortID, path.EndpointA.ChannelID, upgrade, upgrade.Fields.Version)
```
Comment (colin-axner) [inline modules/core/04-channel/keeper/upgrade_test.go:110]: a bit odd this is the only write func embedded into the other function test

File: modules/core/keeper/msg_server.go [hunk 13]
```diff
@@ -756,3 +1132,15 @@ func (k Keeper) UpdateConnectionParams(goCtx context.Context, msg *connectiontyp
 
 	return &connectiontypes.MsgUpdateParamsResponse{}, nil
 }
+
+// UpdateChannelParams defines a rpc handler method for MsgUpdateParams.
+func (k Keeper) UpdateChannelParams(goCtx context.Context, msg *channeltypes.MsgUpdateParams) (*channeltypes.MsgUpdateParamsResponse, error) {
```
Comment (colin-axner) [inline modules/core/keeper/msg_server.go:1137]: I think we are missing a test?

File: modules/core/keeper/msg_server.go [hunk 14]
```diff
@@ -733,6 +734,381 @@ func (k Keeper) Acknowledgement(goCtx context.Context, msg *channeltypes.MsgAckn
...
+	channel, found := k.ChannelKeeper.GetChannel(ctx, msg.PortId, msg.ChannelId)
+	if !found {
+		return nil, errorsmod.Wrapf(channeltypes.ErrChannelNotFound, "port ID (%s) channel ID (%s)", msg.PortId, msg.ChannelId)
+	}
+
+	// if the msgSender is authorized to make and cancel upgrades AND the current channel has not already reached FLUSHCOMPLETE
+	// then we can restore immediately without any additional checks
+	isAuthority := k.GetAuthority() == msg.Signer
+	if isAuthority && channel.State != channeltypes.FLUSHCOMPLETE {
+		k.ChannelKeeper.WriteUpgradeCancelChannel(ctx, msg.PortId, msg.ChannelId, channel.UpgradeSequence)
```
Comment (colin-axner) [inline modules/core/keeper/msg_server.go:1074]: What's the reasoning for not calling:
`k.ChannelKeeper.MustAbortUpgrade(ctx, msg.PortId, msg.ChannelId, err)`?

Comment (colin-axner) [reply modules/core/keeper/msg_server.go:1074]: I think the main diff is just event emission, but makes me wonder if that justifies the separation of code paths

File: modules/core/04-channel/keeper/upgrade.go [hunk 15]
```diff
@@ -0,0 +1,1015 @@
...
+
+// getAbsoluteUpgradeTimeout returns the absolute timeout for the given upgrade.
+func (k Keeper) getAbsoluteUpgradeTimeout(ctx sdk.Context) types.Timeout {
+	upgradeTimeout := k.GetParams(ctx).UpgradeTimeout
+	return types.NewTimeout(clienttypes.ZeroHeight(), uint64(ctx.BlockTime().UnixNano())+upgradeTimeout.Timestamp)
+}
+
+// syncUpgradeSequence ensures current upgrade handshake only continues if both channels are using the same upgrade sequence,
+// otherwise an upgrade error is returned so that an error receipt will be written so that the upgrade handshake may be attempted again with synchronized sequences.
+func (k Keeper) syncUpgradeSequence(ctx sdk.Context, portID, channelID string, channel types.Channel, counterpartyUpgradeSequence uint64) error {
```
[the remark this answers is not in the document]
Comment (damiannolan) [reply modules/core/04-channel/keeper/upgrade.go:845]: good catch! its used in export_test which is doing nothing. 

Let's remove it!

**Review by colin-axner — APPROVED** [review]
Wahoooooooooooooooooooooooooooooo!!!!!!!!!!!!! :rocket: 


Finally! Outstanding, terrific, fantastic, amazing, superb, tremendous, stupendous, excellent, creative, careful, considerate, and diligent work!! Thank you everyone! :heart: :heart: :heart:

File: modules/core/04-channel/keeper/upgrade.go (hunk 11 above)
Comment (damiannolan) [reply modules/core/04-channel/keeper/upgrade.go:985]: I think this func is flagged to removed in another PR

Comment (colin-axner) [reply modules/core/04-channel/keeper/upgrade.go:985]: I think when the channel sequence is told to fast forward by the counterparty (counterparty error receipt is greater than our sequence)?

File: modules/core/04-channel/keeper/upgrade.go (hunk 10 above)
Comment (crodriguezvega) [reply modules/core/04-channel/keeper/upgrade.go:1000]: It will be used [here](https://github.com/cosmos/ibc-go/pull/5405).

Comment (damiannolan) [reply modules/core/04-channel/keeper/upgrade.go:1000]: I think some of these funcs have been left behind as error receipts will be set directly now. 

I think this, as well as `restoreChannel` are supposed to be removed.

File: modules/core/keeper/msg_server.go (hunk 13 above)
Comment (damiannolan) [reply modules/core/keeper/msg_server.go:1137]: looks like it!

File: modules/core/keeper/msg_server.go (hunk 14 above)
Comment (charleenfei) [reply modules/core/keeper/msg_server.go:1074]: i dont think it actually justifies now, considering we moved event emission to this layer (we are currently missing the event emitter in this function but i've fixed that in [this pr](https://github.com/cosmos/ibc-go/pull/5475))

File: modules/apps/transfer/ibc_module_test.go [hunk 16]
```diff
@@ -242,6 +244,224 @@ func (suite *TransferTestSuite) TestOnChanOpenAck() {
...
+	}
+}
+
+func (suite *TransferTestSuite) TestOnChanUpgradeTry() {
+	var (
+		counterpartyUpgrade channeltypes.Upgrade
+		path                *ibctesting.Path
+	)
+
+	testCases := []struct {
```
[the remark this answers is not in the document]
Comment (crodriguezvega) [reply modules/apps/transfer/ibc_module_test.go:324]: Added [here](https://github.com/cosmos/ibc-go/pull/5502).

File: modules/core/04-channel/types/msgs.go [hunk 17]
```diff
@@ -35,6 +36,10 @@ var (
 	_ sdk.HasValidateBasic = (*MsgAcknowledgement)(nil)
 	_ sdk.HasValidateBasic = (*MsgTimeout)(nil)
 	_ sdk.HasValidateBasic = (*MsgTimeoutOnClose)(nil)
+	_ sdk.HasValidateBasic = (*MsgChannelUpgradeInit)(nil)
+	_ sdk.HasValidateBasic = (*MsgChannelUpgradeTry)(nil)
+	_ sdk.HasValidateBasic = (*MsgChannelUpgradeAck)(nil)
+	_ sdk.HasValidateBasic = (*MsgChannelUpgradeConfirm)(nil)
```
[the remark this answers is not in the document]
Comment (crodriguezvega) [reply modules/core/04-channel/types/msgs.go:42]: Added [here](https://github.com/cosmos/ibc-go/pull/5502).

File: modules/core/04-channel/types/msgs.go [hunk 18]
```diff
@@ -374,3 +382,377 @@ func (msg MsgAcknowledgement) ValidateBasic() error {
...
+// NewMsgUpdateChannelParams creates a new instance of MsgUpdateParams.
+func NewMsgUpdateChannelParams(authority string, params Params) *MsgUpdateParams {
+	return &MsgUpdateParams{
+		Authority: authority,
+		Params:    params,
+	}
+}
+
+// ValidateBasic performs basic checks on a MsgUpdateParams.
+func (msg *MsgUpdateParams) ValidateBasic() error {
```
[the remark this answers is not in the document]
Comment (crodriguezvega) [reply modules/core/04-channel/types/msgs.go:721]: Added [here](https://github.com/cosmos/ibc-go/pull/5502).

File: modules/core/keeper/msg_server.go (hunk 13 above)
Comment (crodriguezvega) [reply modules/core/keeper/msg_server.go:1137]: Added [here](https://github.com/cosmos/ibc-go/pull/5494).

File: modules/core/04-channel/keeper/upgrade.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (crodriguezvega) [reply modules/core/04-channel/keeper/upgrade.go:845]: Removed [here](https://github.com/cosmos/ibc-go/pull/5489).

File: modules/apps/29-fee/transfer_test.go [hunk 19]
```diff
@@ -68,3 +68,98 @@ func (suite *FeeTestSuite) TestFeeTransfer() {
...
+					types.NewMsgPayPacketFee(fee, path.EndpointA.ChannelConfig.PortID, path.EndpointA.ChannelID, suite.chainA.SenderAccount.GetAddress().String(), nil),
+					transfertypes.NewMsgTransfer(path.EndpointA.ChannelConfig.PortID, path.EndpointA.ChannelID, ibctesting.TestCoin, suite.chainA.SenderAccount.GetAddress().String(), suite.chainB.SenderAccount.GetAddress().String(), clienttypes.NewHeight(1, 100), 0, ""),
+				}
+
+				res, err := suite.chainA.SendMsgs(msgs...)
+				suite.Require().NoError(err) // message committed
+
+				feeEscrowAddr := suite.chainA.GetSimApp().AccountKeeper.GetModuleAddress(types.ModuleName)
+				escrowBalance := suite.chainA.GetSimApp().BankKeeper.GetBalance(suite.chainA.GetContext(), feeEscrowAddr, sdk.DefaultBondDenom)
+				suite.Require().Equal(escrowBalance.Amount, fee.Total().AmountOf(sdk.DefaultBondDenom))
```
[the remark this answers is not in the document]
Comment (crodriguezvega) [reply modules/apps/29-fee/transfer_test.go:149]: I think it's good to keep the e2e, just to make sure integration with relayer work fine.

File: modules/core/03-connection/keeper/verify.go (hunk 3 above)
Comment (crodriguezvega) [reply modules/core/03-connection/keeper/verify.go:413]: Addressed in https://github.com/cosmos/ibc-go/pull/5532

File: e2e/tests/core/04-channel/upgrades_test.go [hunk 20]
```diff
@@ -0,0 +1,141 @@
...
+		// check the channel version include the fee version
+		version, err := feetypes.MetadataFromVersion(channel.Version)
+		s.Require().NoError(err)
+		s.Require().Equal(feetypes.Version, version.FeeVersion, "the channel version did not include ics29")
+
+		// extra check
+		feeEnabled, err := s.QueryFeeEnabledChannel(ctx, chainB, channelB.PortID, channelB.ChannelID)
+		s.Require().NoError(err)
+		s.Require().Equal(true, feeEnabled)
+	})
```
[the remark this answers is not in the document]
Comment (crodriguezvega) [reply e2e/tests/core/04-channel/upgrades_test.go:118]: Will be addressed [here](https://github.com/cosmos/ibc-go/pull/5539).

File: modules/core/04-channel/keeper/upgrade_test.go [hunk 21]
```diff
@@ -0,0 +1,2892 @@
...
+				path.EndpointB.ChannelConfig.PortID,
+				path.EndpointB.ChannelID,
+				proposedUpgrade,
+				proposedUpgrade.Fields.Version,
+			)
+
+			channel := path.EndpointB.GetChannel()
+			suite.Require().Equal(upgradedChannelEnd, channel)
+
+			upgrade, found := suite.chainB.GetSimApp().IBCKeeper.ChannelKeeper.GetUpgrade(suite.chainB.GetContext(), path.EndpointB.ChannelConfig.PortID, path.EndpointB.ChannelID)
```
[the remark this answers is not in the document]
Comment (crodriguezvega) [reply modules/core/04-channel/keeper/upgrade_test.go:628]: This test calls `WriteUpgradeTryChannel` and the timeout and sequence send are set in `startFlushing`, so I think it's not possible to check either value here.

File: modules/core/04-channel/keeper/upgrade_test.go [hunk 22]
```diff
@@ -0,0 +1,2892 @@
...
+				connectionEnd.State = connectiontypes.UNINITIALIZED
+				path.EndpointA.SetConnection(connectionEnd)
+			},
+			connectiontypes.ErrInvalidConnectionState,
+		},
+		{
+			"invalid counterparty channel state",
+			func() {
+				channel := path.EndpointB.GetChannel()
+				channel.State = types.CLOSED
```
[the remark this answers is not in the document]
Comment (crodriguezvega) [reply modules/core/04-channel/keeper/upgrade_test.go:1296]: Will be addressed in https://github.com/cosmos/ibc-go/pull/5539

File: modules/core/04-channel/types/msgs.go [hunk 23]
```diff
@@ -374,3 +382,377 @@ func (msg MsgAcknowledgement) ValidateBasic() error {
...
+		PortId:            portID,
+		ChannelId:         channelID,
+		ErrorReceipt:      errorReceipt,
+		ProofErrorReceipt: proofErrReceipt,
+		ProofHeight:       proofHeight,
+		Signer:            signer,
+	}
+}
+
+// ValidateBasic implements sdk.Msg
```
[the remark this answers is not in the document]
Comment (crodriguezvega) [reply modules/core/04-channel/types/msgs.go:694]: Addressed in https://github.com/cosmos/ibc-go/pull/5539

File: modules/core/04-channel/keeper/upgrade.go (hunk 9 above)
Comment (colin-axner) [reply modules/core/04-channel/keeper/upgrade.go:735]: #5572

**Pull request merged**